include/jemalloc/internal/rtree_inlines.h

```diff
@@ -465,6 +455,37 @@ rtree_leaf_elm_release(tsdn_t *tsdn, rtree_t *rtree, rtree_leaf_elm_t *elm) {
 }
 
 JEMALLOC_INLINE void
+rtree_szind_slab_update(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
+    uintptr_t key, szind_t szind, bool slab) {
+	assert(!slab || szind < NBINS);
+
+	/*
+	 * The caller implicitly assures that it is the only writer to the szind
+	 * and slab fields.
+	 */
+#ifdef RTREE_LEAF_COMPACT
+	/*
+	 * Another thread may concurrently acquire the elm, which means that
+	 * even though the szind and slab fields will not be concurrently
+	 * modified by another thread, the fact that the lock is embedded in the
+	 * same word requires that a CAS operation be used here.  It would be
+	 * possible to directly CAS here and avoid the lock/unlock operations,
+	 * but at the cost of extra special-purpose code.
+	 */
+	rtree_leaf_elm_t *elm = rtree_leaf_elm_acquire(tsdn, rtree, rtree_ctx,
+	    key, true, false);
+	rtree_leaf_elm_slab_write(tsdn, rtree, elm, true, slab);
+	rtree_leaf_elm_szind_write(tsdn, rtree, elm, true, szind);
+	rtree_leaf_elm_release(tsdn, rtree, elm);
```

Comment:
I'm inclined to say that we should try optimizing this harder. With the current implementation its a CAS with multiple stores. Like you mentioned ideally this should be a single CAS. Probably not a big deal for performance as I won't expect this to be under contention often. But specialized elm_set_atomic function still makes sense to me.